### PR TITLE
Add support for pre-unix-epoch file dates on Apple platforms (#108277)

### PR DESCRIPTION
**Pull request by Byron**
Please note that even though the assertion being hit is the same on MacOS and thus similar to what's described in #108277, on MacOS it's possible to convert the numbers such that they are valid, don't hit the assertion and are round-trippable.
Doing so effectively fixes the issue on Apple platforms.

This PR does not attempt to harden other platforms against negative nanoseconds, which can happen for many reasons including mild filesystem corruption.

----

Time in UNIX system calls counts from the epoch, 1970-01-01. The timespec
struct used in various system calls represents this as a number of seconds and
a number of nanoseconds. Nanoseconds are required to be between 0 and
999_999_999, because the portion outside that range should be represented in
the seconds field; if nanoseconds were larger than 999_999_999, the seconds
field should go up instead.

Suppose you ask for the time 1969-12-31, what time is that? On UNIX systems
that support times before the epoch, that's seconds=-86400, one day before the
epoch. But now, suppose you ask for the time 1969-12-31 23:59:00.1. In other
words, a tenth of a second after one minute before the epoch.  On most UNIX
systems, that's represented as seconds=-60, nanoseconds=100_000_000. The macOS
bug is that it returns seconds=-59, nanoseconds=-900_000_000.

While that's in some sense an accurate description of the time (59.9 seconds
before the epoch), that violates the invariant of the timespec data structure:
nanoseconds must be between 0 and 999999999. This causes this assertion in the
Rust standard library.

So, on macOS, if we get a Timespec value with seconds less than or equal to
zero, and nanoseconds between -999_999_999 and -1 (inclusive), we can add
1_000_000_000 to the nanoseconds and subtract 1 from the seconds, and then
convert.  The resulting timespec value is still accepted by macOS, and when fed
back into the OS, produces the same results. (If you set a file's mtime with
that timestamp, then read it back, you get back the one with negative
nanoseconds again.)

Co-authored-by: Josh Triplett <josh@joshtriplett.org>

### CODE REVIEW

File: library/std/src/sys/unix/time.rs [hunk 1]
```diff
@@ -76,6 +76,18 @@ impl Timespec {
...
+            target_os = "ios",
+            target_os = "tvos",
+            target_os = "watchos"
+        ))]
+        let (tv_sec, tv_nsec) =
+            if (tv_sec <= 0 && tv_sec > i64::MIN) && (tv_nsec < 0 && tv_nsec > -1_000_000_000) {
+                (tv_sec - 1, tv_nsec + 1_000_000_000)
+            } else {
+                (tv_sec, tv_nsec)
+            };
```
Comment (Urgau) [inline library/std/src/sys/unix/time.rs:90]: This should have a comment explaining what is going on here. Something similar to the PR explanation.

File: library/std/src/sys/unix/time.rs [hunk 2]
```diff
@@ -76,6 +76,18 @@ impl Timespec {
     }
 
     const fn new(tv_sec: i64, tv_nsec: i64) -> Timespec {
+        #[cfg(any(
```
Comment (joshtriplett) [inline library/std/src/sys/unix/time.rs:79]:
```suggestion
        // Apple systems return times before the epoch with negative nanoseconds.
        // Borrow from the seconds field to keep the nanoseconds in range.
        #[cfg(any(
```

File: library/std/src/sys/unix/time.rs (hunk 1 above)
Comment (Byron) [reply library/std/src/sys/unix/time.rs:102]: Thanks for the hint!

Here is the added comment in clear text:

```
On Apple OS, dates before epoch are represented differently than on other
Unix platforms: e.g. 1/10th of a second before epoch is represented as `seconds=-1`
and `nanoseconds=100_000_000` on other platforms, but is `seconds=0` and
`nanoseconds=-900_000_000` on Apple OS.

To compensate, we first detect this special case by checking if both
seconds and nanoseconds are in range, and then correct the value for seconds
and nanoseconds to match the common unix representation.

Please note that Apple OS nonetheless accepts the standard unix format when
setting file times, which makes this compensation round-trippable and generally
transparent.
```

It's a bit wordy, but complete, and @joshtriplett provided a more concise version of it as well which might be more desirable after all. That very much depends on what has proven to be desirable in the standard library, I am happy to make adjustments as needed.
